packages/@n8n/api-types/src/dto/folders/list-folder-query.dto.ts

```diff
@@ -0,0 +1,181 @@
+import { ApplicationError, jsonParse } from 'n8n-workflow';
+import { z } from 'zod';
+import { Z } from 'zod-class';
+
+const VALID_SELECT_FIELDS = [
+	'id',
+	'name',
+	'createdAt',
+	'updatedAt',
+	'project',
+	'tags',
+	'parentFolder',
+	'workflowCount',
+] as const;
+
+const VALID_SORT_OPTIONS = [
+	'name:asc',
+	'name:desc',
+	'createdAt:asc',
+	'createdAt:desc',
+	'updatedAt:asc',
+	'updatedAt:desc',
+] as const;
+
+// Filter schema - only allow specific properties
+export const filterSchema = z
+	.object({
+		parentFolderId: z.string().optional(),
+		name: z.string().optional(),
+		tags: z.array(z.string()).optional(),
+	})
+	.strict();
+
+// Common transformers
+const parseJsonArray = (val: string): unknown => {
+	if (!val.trim().startsWith('[')) {
+		throw new ApplicationError('Expected a JSON array starting with [');
+	}
+	return jsonParse(val);
+};
+
+const isArrayOfStrings = (val: unknown): val is string[] => {
+	return Array.isArray(val) && val.every((item) => typeof item === 'string');
+};
+
+// ---------------------
+// Parameter Validators
+// ---------------------
+
+// Filter parameter validation
+const filterValidator = z
+	.string()
+	.optional()
+	.superRefine((val, ctx) => {
+		if (!val) return;
+
+		try {
+			const parsed = jsonParse(val);
+			try {
+				filterSchema.parse(parsed);
+			} catch (e) {
+				ctx.addIssue({
+					code: z.ZodIssueCode.custom,
+					message: 'Invalid filter fields',
+					path: ['filter'],
+				});
+				return z.NEVER;
+			}
+		} catch (e) {
+			ctx.addIssue({
+				code: z.ZodIssueCode.custom,
+				message: 'Invalid filter format',
+				path: ['filter'],
+			});
+			return z.NEVER;
+		}
+	})
+	.transform((val) => {
+		if (!val) return undefined;
+		return filterSchema.parse(jsonParse(val));
+	});
+
+// Skip parameter validation
+const skipValidator = z
+	.string()
+	.optional()
+	.transform((val) => (val ? parseInt(val, 10) : 0))
+	.refine((val) => val === undefined || !isNaN(val), {
+		message: 'Skip must be a valid number',
+	});
```

Comment:
refine doesn't need to check for undefined here

```suggestion
	.refine((val) => !isNaN(val), {
		message: 'Skip must be a valid number',
	});
```